test/e2e/clusterclass_changes.go

```diff
@@ -0,0 +1,403 @@
+	})
+}
+
+// modifyClusterClassControlPlaneAndWaitInput is the input type for modifyControlPlaneViaClusterClassAndWait.
+type modifyClusterClassControlPlaneAndWaitInput struct {
+	ClusterProxy             framework.ClusterProxy
+	ClusterClass             *clusterv1.ClusterClass
+	Cluster                  *clusterv1.Cluster
+	ModifyControlPlaneFields map[string]interface{}
+	WaitForControlPlane      []interface{}
+}
+
+// modifyControlPlaneViaClusterClassAndWait modifies the ControlPlaneTemplate of a ClusterClass by setting ModifyControlPlaneFields
+// and waits until the changes are rolled out to the ControlPlane of the Cluster.
+func modifyControlPlaneViaClusterClassAndWait(ctx context.Context, input modifyClusterClassControlPlaneAndWaitInput) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for ModifyControlPlaneViaClusterClassAndWait")
+	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling ModifyControlPlaneViaClusterClassAndWait")
+	Expect(input.ClusterClass).ToNot(BeNil(), "Invalid argument. input.ClusterClass can't be nil when calling ModifyControlPlaneViaClusterClassAndWait")
+	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling ModifyControlPlaneViaClusterClassAndWait")
+
+	mgmtClient := input.ClusterProxy.GetClient()
+
+	log.Logf("Modifying the ControlPlaneTemplate of ClusterClass %s/%s", input.ClusterClass.Namespace, input.ClusterClass.Name)
+
+	// Get ControlPlaneTemplate object.
+	controlPlaneTemplateRef := input.ClusterClass.Spec.ControlPlane.Ref
+	controlPlaneTemplate, err := external.Get(ctx, mgmtClient, controlPlaneTemplateRef, input.ClusterClass.Namespace)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Create a new ControlPlaneTemplate object with a new name and ModifyControlPlaneFields set.
+	newControlPlaneTemplate := controlPlaneTemplate.DeepCopy()
+	newControlPlaneTemplateName := fmt.Sprintf("%s-%s", controlPlaneTemplateRef.Name, util.RandomString(6))
+	newControlPlaneTemplate.SetName(newControlPlaneTemplateName)
+	newControlPlaneTemplate.SetResourceVersion("")
+	for fieldPath, value := range input.ModifyControlPlaneFields {
+		templateFieldPath := append([]string{"spec", "template"}, strings.Split(fieldPath, ".")...)
+		Expect(unstructured.SetNestedField(newControlPlaneTemplate.Object, value, templateFieldPath...)).To(Succeed())
+	}
+	Expect(mgmtClient.Create(ctx, newControlPlaneTemplate)).To(Succeed())
+
+	// Patch the ClusterClass ControlPlaneTemplate ref to reference the new ControlPlaneTemplate.
+	patchHelper, err := patch.NewHelper(input.ClusterClass, mgmtClient)
+	Expect(err).ToNot(HaveOccurred())
+	controlPlaneTemplateRef.Name = newControlPlaneTemplateName
+	Expect(patchHelper.Patch(ctx, input.ClusterClass)).To(Succeed())
+
+	log.Logf("Waiting for ControlPlane rollout to complete.")
```

Comment:
Q: if I got this right we are not waiting for the rollout to complete, but only waiting for the control plane object spec to be modified. Should we wait for CP machines rollout instead (might be this can easily be achieved by looked at upgraded replicas in status)

Reply:
My idea was to just check that it's rolled out to the ControlPlane because the ControlPlane rollout is the concern of the ControlPlane provider (and we have other e2e tests covering this)

It also speeds up the test a bit because we don't have to wait that long.

Note: this test can currently also be used with other control plane providers which don't support machines.

Reply:
No problem. I should have realized that dependency :). I'll adjust accordingly